text/044-vo-named-format.md

```diff
@@ -0,0 +1,163 @@
+and position information. In particular, the contents of a segment is uniquely
+determined by its position in the list of a segment. As of 8.11, the segments
+are, in the file order:
+- Summary of the file
+- Logical contents of the library
+- Delayed universes (occurs in delayed compilation of partially compiled files)
+- Delayed proofs (only relevant for partially compiled files)
+- Opaque proofs
+
+This representation is very rigid. Most notably, adding optional segments
+requires changing all tools manipulating vo files. For instance, the delayed
+proof segment is non-empty only in `.vio` files, but all callers need to be
+aware of it. Another issue is that it prevents having a variable number of
+segments, e.g. all opaque proofs need to be stored in the same OCaml array and
+cannot be stored in a per-proof basis, since the number of opaque proofs
+depends on each file. In turn, this does not allow clever on-disk storing
+schemes of opaque proofs as all proofs need to be loaded in memory in the array
+when marshalling this segment.
+
+Instead, we propose a COFF where segments can be accessed by name instead of
+position. This would solve both problems at once, by making the format at the
+same time more robust and more extensible.
+
+# Detailed design
+
+## Inspiration
+
+We introduce a new COFF inspired by the ZIP data format, following experiments
+with the `camlzip` library. We decided against ZIP because this format suffers
+from historical 16-bit and 32-bit limitations, e.g. a ZIP file can only contain
+up to 65535 subfiles, and weight up to about 4GB. Working around these
+limitations requires the introduction of the ZIP64 variant which is very
+inelegant in order to preserve backwards compatibility with ZIP to some extent.
+The ZIP format also suffers from a few additional quirks (e.g. ambiguities in
+the serialization scheme) that render it improper for high-confidence formats.
+
+## On compression
+
+The experiments also showed that systematic compression would incur an important
+runtime cost. ZIP supports several compression schemes, of which `camlzip`
+implements two, `DEFLATE` (LZ-style, with various levels of compression) and
+`STORE` (no compression). All levels of `DEFLATE` triggered important slowdown,
+while `STORE` was invisible. Further profiling showed that both compression and
+decompression were costly, as Coq typically loads a lot of files at once by
+transitive dependency.
+
+For this reason, the new COFF we propose does not implement any compression
+and merely stores the data as-is.
+
+## Overview
+
+The major difference introduced by the new COFF is a metadata registry at the
+end of the file that contains information about the segments, and in particular
+their names. Merely parsing this metadata is enough to know the position of
+a segment inside the file. The position of this registry is known by parsing
+the header of the file.
+
+Another design choice is the reliance on 64-bit integers to indicate file
```

Comment:
@corwin-of-amber , how would this impact jsCoq ? Also, quite a few number of users use 32bit Coq as it  does provide in general a non-trivial speedup; we should measure the impact here.

We could allow both 31 / 63 bits and add a tag to the file; but indeed I'm split on that.

Reply:
Well: if there are arbitrary Int64s in the data structure, and `Marshal` is used to serialize it, then we will have errors when trying to save it with `Compat_32` if the values are too large.

Reply:
If that's the only problem I think we are good for now; I understand values a file offsets so if we are using `.vo` files > 4GiB then we have other things to worry about I think.

Reply:
Yeah, I think we can safely assume that we would not want to distribute such large files over the Web anyway...

Reply:
I don't know for JSCoq (even though I suspect it's the same), but it shouldn't impact *at all* the standard OCaml runtime in terms of performance. This is merely allocating a handful of words while the deserialization itself can be arbitrarily large. This is really just about being able to seek in big files, it doesn't affect the serialization of the data proper.

Reply:
@ppedrot javascript doesn't have 64bit integers O_O

Reply:
FWIW, WebAssembly, while being a 32-bit platform, has 64-bit integers. I think that, at least for the sake of interop with WebAssembly, they will be introduced into JavaScript as well, eventually.

Reply:
@ejgallego How is JScoq interpreting the `int64` type? The C code is not bound to a software-emulated implementation?

Reply:
Not exactly – js_of_ocaml translates OCaml to plain JavaScript, there is no "virtual machine" other than the JS interpreter. I believe that if the type is defined explicitly as `int64` (rather than `int`), then it *will* be represented reliably using a tuple of smaller integers. I have not tried to use `int64` in this context, though.